qualtran/bloqs/factoring/mod_mul.py

```diff
@@ -104,6 +116,94 @@ def wire_symbol(self, soq: 'Soquet') -> 'WireSymbol':
             return directional_text_box(f'*={self.k}', side=soq.reg.side)
 
 
+@frozen
+class MontgomeryModDbl(Bloq):
+    r"""An n-bit modular doubling gate.
+
+    This gate is designed to operate on integers in the Montgomery form.
+    Implements |x> => |2 * x % p> using $2n$ Toffoli gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+        p: The modulus for the doubling.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585) Fig 6d and 8
+    """
+
+    bitsize: int
+    p: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', QMontgomeryUInt(self.bitsize))])
+
+    def on_classical_vals(self, x: 'ClassicalValT') -> Dict[str, 'ClassicalValT']:
+
+        x *= 2
+        x -= self.p
+
+        if x < 0:
+            x += self.p
```

Comment:
is this right? why can't we just do `% self.p`?

Reply:
If we assume the input value is given in Montgomery Form `x % self.p` would be correct still, however I just tried to match the actual bloq implementation as closely as possible. Open to whatever you'd prefer. If the number is not in Montgomery Form then that equation would not hold.

Reply:
Philosophically: this method should give the desired behavior in a readable form. It's the "reference implementation" for a unit test. 

To get the mechanistic behavior, you can do `bloq.decompose_bloq().call_classically`. Comparing these two are indeed our favored unit test